src/Netscript/APIWrapper.ts

```diff
@@ -0,0 +1,157 @@
+import type { WorkerScript } from "./WorkerScript";
+import { makeRuntimeRejectMsg } from "../NetscriptEvaluator";
+import { Player } from "../Player";
+
+type ExternalFunction = (...args: any[]) => any;
+type ExternalAPI = {
+  [string: string]: ExternalAPI | ExternalFunction;
+};
+
+type InternalFunction<F extends (...args: unknown[]) => unknown> = (
+  ctx: NetscriptContext,
+) => (...args: unknown[]) => ReturnType<F>;
+export type InternalAPI<API> = {
+  [Property in keyof API]: API[Property] extends ExternalFunction
+    ? InternalFunction<API[Property]>
+    : API[Property] extends ExternalAPI
+    ? InternalAPI<API[Property]>
+    : never;
+};
+
+type WrappedNetscriptFunction = (...args: unknown[]) => unknown;
+type WrappedNetscriptAPI = {
+  [string: string]: WrappedNetscriptAPI | WrappedNetscriptFunction;
+};
+
+export type NetscriptContext = {
+  makeRuntimeErrorMsg: (message: string) => string;
+  log: (message: () => string) => void;
+  workerScript: WorkerScript;
+  function: string;
+  helper: WrappedNetscriptHelpers;
+};
+
+type NetscriptHelpers = {
+  updateDynamicRam: (fnName: string, ramCost: number) => void;
+  makeRuntimeErrorMsg: (caller: string, msg: string) => string;
+  string: (funcName: string, argName: string, v: unknown) => string;
+  number: (funcName: string, argName: string, v: unknown) => number;
+  boolean: (v: unknown) => boolean;
+  getServer: (hostname: string, callingFnName: string) => BaseServer;
+  checkSingularityAccess: (func: string) => void;
+  hack: (hostname: any, manual: any, { threads: requestedThreads, stock }?: any) => Promise<number>;
+  getValidPort: (funcName: string, port: any) => IPort;
+};
+
+type WrappedNetscriptHelpers = {
+  updateDynamicRam: (ramCost: number) => void;
+  makeRuntimeErrorMsg: (msg: string) => string;
+  string: (argName: string, v: unknown) => string;
+  number: (argName: string, v: unknown) => number;
+  boolean: (v: unknown) => boolean;
+  getServer: (hostname: string) => BaseServer;
+  checkSingularityAccess: () => void;
+  hack: (hostname: any, manual: any, { threads: requestedThreads, stock }?: any) => Promise<number>;
+  getValidPort: (port: any) => IPort;
+};
+
+function wrapFunction(
+  helpers: NetscriptHelpers,
+  wrappedAPI: any,
+  workerScript: WorkerScript,
+  func: (_ctx: NetscriptContext) => (...args: unknown[]) => unknown,
+  ...tree: string[]
+): void {
+  const functionPath = tree.join(".");
+  const functionName = tree.pop();
+  if (typeof functionName !== "string") {
+    throw makeRuntimeRejectMsg(workerScript, "Failure occured while wrapping netscript api");
+  }
+  const ctx = {
+    makeRuntimeErrorMsg: (message: string) => {
+      return helpers.makeRuntimeErrorMsg(functionPath, message);
+    },
+    log: (message: () => string) => {
+      workerScript.log(functionPath, message);
+    },
+    workerScript,
+    function: functionName,
+    helper: {
+      updateDynamicRam: (ramCost: number) => helpers.updateDynamicRam(functionName, ramCost),
```

Comment:
I think this should be `updateDynamicRam(functionPath` ?

Reply:
oh nvm that is not how the current implementation works ... but should it be?

Reply:
thinking about it since the wrapper calls updateDynamicRam itself theres no reason to include it in the helpers in context
so il actually just remove that